### PR TITLE
Introduce government data alert

### PR DESCRIPTION
**Pull request by JonathanHallam**
Trello - https://trello.com/c/FaHAEajH/1318-alert-2ndline-when-we-have-no-cached-data-on-governments

This iterates out healthcheck endpoint to include a check for government data, a connection to redis and to the database. This will allow us to alert 2nd line and our slack channel in the event that government data is unavailable. 

### Relies on
https://github.com/alphagov/govuk-developer-docs/pull/2301
https://github.com/alphagov/govuk-puppet/pull/10129

### CODE REVIEW

**Comment by cbaines** [comment]
A more general comment, but with the Email Alert API, the Platform Health Team has been working on trying to get the per-machine healthcheck back to just being about the health of the app on that machine [1].

1: https://github.com/alphagov/email-alert-api/commits/master/app/controllers/healthcheck_controller.rb

At least in the AWS world, the ELB's can use healthchecks to decide which machines to route traffic to. At least currently for the Email Alert API, it's not looking at the status of the healthcheck, but that would be good to change.

Regarding the `Healthcheck::GovernmentDataCheck` aspect of the health check, is this something that can differ between machines?

**Comment by kevindew** [comment]
Thanks @cbaines for the input - this would be a check that applies to all machines

Are there decisions (or better documentation) on what we aspire to have on /healthcheck and for things that aren't covered there how. I gather for Email Alert API it's moved checks out of using Ruby and instead to monitor graphite metrics, which makes me unsure of precedent to follow for Ruby based checks if they don't use /healthcheck.

**Comment by cbaines** [comment]
> Are there decisions (or better documentation) on what we aspire to have on /healthcheck and for things that aren't covered there how. I gather for Email Alert API it's moved checks out of using Ruby and instead to monitor graphite metrics, which makes me unsure of precedent to follow for Ruby based checks if they don't use /healthcheck.

I don't think anyone has made any decisions or documentation. 

I guess here, I don't see much significance in where the check is happening and what language it's implemented in. For example, you could have Content Publisher run a Sidekiq job each minute that computes the age of the government repository data, and pushes that value to Graphite (through statsd). You could then have a Icinga check that looks at that value, and checks if it's over 6 hours. Even though it's Icinga doing the checking, I think it's still possible to do what I'm guessing you want to. Also, you get the added benefit of being able to look at the trend in the data through Graphite.

**Comment by kevindew** [comment]
> I don't think anyone has made any decisions or documentation.

I see, hmm maybe we need PH to do some thinking about purpose of this endpoint. It looks a bit like Email Alert API has gone in an opinionated direction of its own that leaves a few questions/discussions. Maybe an RFC might be the way to go to get comments and documentation at the same time.


> I guess here, I don't see much significance in where the check is happening and what language it's implemented in. For example, you could have Content Publisher run a Sidekiq job each minute that computes the age of the government repository data, and pushes that value to Graphite (through statsd). You could then have a Icinga check that looks at that value, and checks if it's over 6 hours. Even though it's Icinga doing the checking, I think it's still possible to do what I'm guessing you want to. Also, you get the added benefit of being able to look at the trend in the data through Graphite.

Well I think the things that lost there is the ease of the check. We take something that's a quick Ruby method and replace it with logic split across two repos (puppet and here) and then need an additional asynchronous job process. That feels a bit cumbersome in comparison :-/ 

If we have two purposes for the endpoint that we're trying to break up maybe we need two endpoints? Or if we want to be regularly updating graphite for lots of apps (I assume many are in similar boat) maybe we need something in govuk_app_config that provides common functionality.

**Comment by cbaines** [comment]
> > I guess here, I don't see much significance in where the check is happening and what language it's implemented in. For example, you could have Content Publisher run a Sidekiq job each minute that computes the age of the government repository data, and pushes that value to Graphite (through statsd). You could then have a Icinga check that looks at that value, and checks if it's over 6 hours. Even though it's Icinga doing the checking, I think it's still possible to do what I'm guessing you want to. Also, you get the added benefit of being able to look at the trend in the data through Graphite.
> 
> Well I think the things that lost there is the ease of the check. We take something that's a quick Ruby method and replace it with logic split across two repos (puppet and here) and then need an additional asynchronous job process. That feels a bit cumbersome in comparison :-/
> 
> If we have two purposes for the endpoint that we're trying to break up maybe we need two endpoints? Or if we want to be regularly updating graphite for lots of apps (I assume many are in similar boat) maybe we need something in govuk_app_config that provides common functionality.

I can see us ending up with some kind of `/metrics` endpoint at some point (Prometheus...), but at least for now with statsd and Graphite, just fitting the code in where relevant seems OK to me.

Anyway, I don't see anything wrong with what's being done here, the only thing to avoid is the healthcheck timing out unnecessarily, and it doesn't look like there's a possibility of that here. I just wanted to make you aware of what had been happening elsewhere on GOV.UK.

**Comment by kevindew** [comment]
> I can see us ending up with some kind of `/metrics` endpoint at some point (Prometheus...), but at least for now with statsd and Graphite, just fitting the code in where relevant seems OK to me.

Okey doke 👍 
 
> Anyway, I don't see anything wrong with what's being done here, the only thing to avoid is the healthcheck timing out unnecessarily, and it doesn't look like there's a possibility of that here. I just wanted to make you aware of what had been happening elsewhere on GOV.UK.

Ah sure that's cool, I think I misinterpreted a bit that we were on a new direction. Apologies if I came across like that. Thanks for the comms on it, it's really helpful to having your teams insight too.

File: lib/bulk_data/cache.rb [hunk 1]
```diff
@@ -22,6 +22,10 @@ def self.read(key)
       cache.read(key) || (raise NoEntryError)
     end
 
+    def self.populated?(key)
+      cache.exist?(key)
+    end
```
Comment (kevindew) [inline lib/bulk_data/cache.rb]: Rather than explicitly adding a method I think it'd be preferable to delegate `exist?` to the cache instance: https://github.com/alphagov/content-publisher/blob/c14461000856ef638303eb63b1e4315d638e13ec/lib/bulk_data/cache.rb#L11

I think we also can kill of the clear method as that is already delegated.

File: lib/bulk_data/government_repository.rb [hunk 2]
```diff
@@ -46,5 +46,13 @@ def populate_cache(older_than: nil)
     rescue GdsApi::BaseError
       raise RemoteDataUnavailableError
     end
+
+    def age_of_cache
+      (Time.zone.now - Cache.written_at(CACHE_KEY)).seconds
```
Comment (kevindew) [inline lib/bulk_data/government_repository.rb]: Time.current is preferred to using Time.zone.now, I wonder if I can make that a linter rule as it comes up a bit 🤔 

I'm not sure we need the `.seconds` as I believe this will execute the same with it or without it.

File: lib/healthcheck/government_data_check.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+module Healthcheck
+  class GovernmentDataCheck
+    def name
+      :government_data_check
+    end
+
+    def status
+      return GovukHealthcheck::CRITICAL unless government_data_availible?
+
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: It feels a bit unusual to have new lines between the guard clauses.

File: lib/healthcheck/government_data_check.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    def enabled?
+      true
+    end
+
+  private
+
+    def government_data_availible?
+      BulkData::GovernmentRepository.new.cache_populated?
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: It'd make sense to have the repository as an instance variable rather than creating a new instance each time. 

This would also make the code shorter so you could probably kill this method and instead have the guard clause be: `return GovukHealthcheck::CRITICAL unless repository.cache_populated?`

File: lib/healthcheck/government_data_check.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    def data_stale?
+      repository = BulkData::GovernmentRepository.new
+      repository.age_of_cache > 6.hours
+    end
+
+    def warning_details_content
+      data_age = BulkData::GovernmentRepository.new.age_of_cache
+      "Government data not refreshed in #{data_age / 3600} hours #{data_age / 60} minutes"
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: This doesn't look correct - and also seems to be missing a test - as if you had 7200 as the age you'd get an output of "Government data not refreshed in 2 hours 120 minutes". You'll probably want to pull out the modulus operator.

Also using ActiveSupport::Duration and division for conversion makes some pretty strange intermediary numbers for example:
```
irb(main):034:0> 7200.seconds / 3600
=> 2 seconds
```

So might be actually be a not very friendly class to use and better to use an integer.

File: spec/lib/healthcheck/government_data_check_spec.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    context "government data is unavailable" do
+      it "returns critical" do
+        gov_data_check = Healthcheck::GovernmentDataCheck.new
+        expect(gov_data_check.status).to be(:critical)
+      end
+    end
+
+    context "government data is set to expire" do
+      it "gives a warning" do
+        BulkData::Cache.write("government-v1", "value")
```
Comment (kevindew) [inline spec/lib/healthcheck/government_data_check_spec.rb]: The use of this id makes an unnecessary coupling between this test and the constant in GovernmentRepository. I'd suggest instead using the helper instead: https://github.com/alphagov/content-publisher/blob/c074cd3ebb6c885f1fa3d7112cd7af6bbb4d01da/spec/support/bulk_data_helper.rb#L4

File: lib/bulk_data/government_repository.rb (hunk 2 above)
Comment (JonathanHallam) [reply lib/bulk_data/government_repository.rb]: Absolutely right on the `.seconds`. I'm happy to do the linter rule. Do you think that needs doing just in content publisher or in https://github.com/alphagov/rubocop-govuk?

File: spec/lib/healthcheck/government_data_check_spec.rb (hunk 6 above)
Comment (JonathanHallam) [reply spec/lib/healthcheck/government_data_check_spec.rb]: Nice one, forgot that you could do this, have adjusted.

File: lib/healthcheck/government_data_check.rb (hunk 5 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Rookie error on this one, have adjusted the output and written a test for it. I've left the `cache_age` method returning an `ActiveSupport::Duration` and turned that into an integer on the `warning_details_content` method. It means the return clauses above look a bit nicer as I can check for `@government_repo.cache_age > 6.hours` rather than `@government_repo.cache_age > 21600`

File: lib/bulk_data/government_repository.rb [hunk 7]
```diff
@@ -46,5 +46,13 @@ def populate_cache(older_than: nil)
     rescue GdsApi::BaseError
       raise RemoteDataUnavailableError
     end
+
+    def age_of_cache
```
[the remark this answers is not in the document]
Comment (JonathanHallam) [reply lib/bulk_data/government_repository.rb]: done :)

File: lib/bulk_data/cache.rb (hunk 1 above)
Comment (JonathanHallam) [reply lib/bulk_data/cache.rb]: clear method is now gone :)

File: lib/healthcheck/government_data_check.rb (hunk 3 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: yep, sorry, had to leave one of them in because the linter complained though

File: lib/healthcheck/government_data_check.rb (hunk 4 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Done, that really cleared it up, thanks

File: db/schema.rb [hunk 8]
```diff
@@ -371,10 +371,10 @@
   create_table "whitehall_migrations", force: :cascade do |t|
     t.text "organisation_content_id"
     t.text "document_type"
-    t.text "document_subtypes", array: true
     t.datetime "finished_at"
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
+    t.text "document_subtypes", array: true
```
Comment (kevindew) [inline db/schema.rb]: Looks like this has slipped in on an unrelated commit?

File: lib/healthcheck/government_data_check.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    def initialize
+      @government_repo = BulkData::GovernmentRepository.new
+    end
+
+    def name
+      :government_data_check
+    end
+
+    def status
+      return GovukHealthcheck::CRITICAL unless @government_repo.cache_populated?
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: Convention would be to use an attribute reader for government_repo

File: lib/healthcheck/government_data_check.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+    def enabled?
+      true
+    end
+
+  private
+
+    def warning_details_content
+      data_age = @government_repo.cache_age.to_i
+      "Government data not refreshed in #{data_age / 3600} hours #{data_age % 60} minutes"
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: The arithmetic is still wrong here.

For example if you have 4000 seconds it is 1 hours 6 minutes. Whereas this will say "Government data not refreshed in 1 hours 40 minutes" - which also reveals plural issues.

File: spec/lib/healthcheck/government_data_check_spec.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    end
+  end
+
+  describe "#details" do
+    let(:gov_data_check) { described_class.new }
+
+    it "displays appropriate messages" do
+      government_repo = BulkData::GovernmentRepository.new
+      gov_data_check.instance_variable_set(:@government_repo, government_repo)
+      allow(government_repo).to receive(:cache_age).and_return(1.second)
```
Comment (kevindew) [inline spec/lib/healthcheck/government_data_check_spec.rb]: I think this now returns an integer

File: spec/lib/healthcheck/government_data_check_spec.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+      expect(gov_data_check.details[:warning]).to eq("Government data not refreshed in 0 hours 1 minutes")
```
Comment (kevindew) [inline spec/lib/healthcheck/government_data_check_spec.rb]: Should 1 second really output 1 minute?

File: lib/tasks/populate_government_data.rake [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+namespace :app do
+  desc "Populate government data"
+  task populate_government_data: :environment do
+    PopulateBulkDataJob.perform_now
+  end
+end
```
Comment (kevindew) [inline lib/tasks/populate_government_data.rake]: I can see a few issues here (namespace app, task focused on government data but the job named differently) though mostly I'm wondering why we need the task since it's one line, may as well suggest someone runs the command in the console.

File: spec/requests/healthcheck_spec.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    context "when the healthchecks pass" do
+      before do
+        populate_default_government_bulk_data
+      end
+
+      it "returns a status of 'ok'" do
+        get healthcheck_path
+        expect(JSON.parse(response.body)["status"]).to eq("ok")
+      end
+    end
```
Comment (kevindew) [inline spec/requests/healthcheck_spec.rb]: I don't think this request needs to be concerned about what the statuses are - since there is no logic in the controller about what is returned just which tests are run. 

I think you'd be fine to test that the healthcheck does two things:

1) returns a 200 HTTP status
2) returns JSON with a status attribute (doesn't matter what the value is)

File: db/schema.rb (hunk 8 above)
Comment (JonathanHallam) [reply db/schema.rb]: Honestly have no idea how this got in, I've not had to look at this :/ at any rate I've removed it.

File: lib/healthcheck/government_data_check.rb (hunk 9 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: I've done this, can I ask why this is convention though?

File: lib/healthcheck/government_data_check.rb (hunk 10 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Changed the working of this after a face to face conversation. Decided that it was accurate enough to just have hours given that this won't show up unless the data has been missing for more than 6.

File: spec/lib/healthcheck/government_data_check_spec.rb (hunk 11 above)
Comment (JonathanHallam) [reply spec/lib/healthcheck/government_data_check_spec.rb]: Sorted

File: spec/lib/healthcheck/government_data_check_spec.rb (hunk 12 above)
Comment (JonathanHallam) [reply spec/lib/healthcheck/government_data_check_spec.rb]: Have now dropped this entirely after face to face conversation.

File: lib/tasks/populate_government_data.rake (hunk 13 above)
Comment (JonathanHallam) [reply lib/tasks/populate_government_data.rake]: Have removed

File: spec/requests/healthcheck_spec.rb (hunk 14 above)
Comment (JonathanHallam) [reply spec/requests/healthcheck_spec.rb]: Nice, I've changed the tests and removed the context block now that it doesn't make sense

File: lib/healthcheck/government_data_check.rb (hunk 9 above)
Comment (kevindew) [reply lib/healthcheck/government_data_check.rb]: Mostly just what GDS went with when it started, which will then be from it being a very common Ruby convention.

There's a bunch of arguments why attr_reader is the preferred approach online: https://github.com/rubocop-hq/ruby-style-guide/issues/538 (which proposed the other approach), https://ivoanjo.me/blog/2017/09/20/why-i-always-use-attr_reader-to-access-instance-variables/ but I don't think any are earth shattering reasons. For me I'm just all for consistency, which trumps my own minor preference for attr_reader.

File: lib/healthcheck/government_data_check.rb [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+      @government_repo = BulkData::GovernmentRepository.new
+    end
+
+    def name
+      :government_data_check
+    end
+
+    def status
+      return GovukHealthcheck::CRITICAL unless government_repo.cache_populated?
+      return GovukHealthcheck::WARNING if government_repo.cache_age > 21600
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: This can still be 6.hours:

```
irb(main):001:0> 3700 > 1.hour
=> true
irb(main):002:0> 3500 > 1.hour
=> false
```

File: lib/healthcheck/government_data_check.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+    def enabled?
+      true
+    end
+
+  private
+
+    def warning_details_content
+      data_age = government_repo.cache_age.to_i
+      "Government data not refreshed in #{data_age / 3600} hours."
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: If we're changing integers we could change the 3600 to 1.hour not sure if that's much of a reading win or not?

File: spec/lib/healthcheck/government_data_check_spec.rb [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+      end
+    end
+  end
+
+  describe "#details" do
+    let(:gov_data_check) { described_class.new }
+
+    it "displays appropriate messages" do
+      government_repo = BulkData::GovernmentRepository.new
+      gov_data_check.instance_variable_set(:@government_repo, government_repo)
```
Comment (kevindew) [inline spec/lib/healthcheck/government_data_check_spec.rb]: This seems a quite unusual mocking approach. A more common one would be expect to receive new and return instance double. For example: https://github.com/alphagov/content-publisher/blob/e06613a8a3de96245ec7055e362a6dae6da5775f/spec/services/resync_document_service_spec.rb#L115-L118

However I think you do better continuing the approach you did earlier in this test and using the `travel_to` method

File: lib/healthcheck/government_data_check.rb [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+      return GovukHealthcheck::WARNING if government_repo.cache_age > 21600
+
+      GovukHealthcheck::OK
+    end
+
+    def details
+      {
+        critical: "No government data availible",
+        warning: warning_details_content,
+      }
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: Sorry I missed this before. Is it normal that these methods will return the critical and warning messages even if the healthcheck is not in a critical or warning state? It feels like that is information that'd show in icinga.

Having a look around I'm not too sure what the purpose and structure of details needs to be. It seems common to have a value, total and thresholds (https://github.com/alphagov/govuk_app_config/blob/master/lib/govuk_app_config/govuk_healthcheck/threshold_check.rb#L23-L32) though I don't know what icinga parses or not.

File: spec/requests/healthcheck_spec.rb [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+RSpec.describe "Healthcheck" do
+  describe "GET /healthcheck" do
+    before do
+      populate_default_government_bulk_data
+    end
+
+    it "returns a 200 HTTP status" do
+      get healthcheck_path
+      expect(response.status).to eq(200)
```
Comment (kevindew) [inline spec/requests/healthcheck_spec.rb]: minor, we've done this elsewhere with `expect(response).to have_http_status(:ok)`

File: lib/healthcheck/government_data_check.rb (hunk 18 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: I could have sworn that I'd seen it done like this in email alert api but I can't find it now. It does seem to be how [DGU does it](https://github.com/alphagov/datagovuk_publish/blob/a8468e62cb94ff755215b9cf23da345c5f4c31bc/app/models/healthcheck/recurring_jobs/ckan_org_sync.rb#L25) but you're right at this doesn't look entirely normal. I'll have a crack at it.

Comment (kevindew) [reply lib/healthcheck/government_data_check.rb]: Sure, they just ripped loads out of Email Alert API so maybe can find them in the past, though it'd be best to have a current one so we can look at it in icinga.

Very confusing 🤔

Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: ah found it https://github.com/alphagov/email-alert-api/commit/0c17e757179cb4b0308bf0b49bb3b19842c15bc2#diff-9fccbed8f77bf59ae535edc077ce5e21

Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: doesn't mean that it worked but I'm glad I wasn't losing my marbles. I might test it in integration this afternoon, as it stands visiting `/healthcheck` will show both results but I don't know what it'll do in icinga so I'll be able to see what actually works.

Comment (kevindew) [reply lib/healthcheck/government_data_check.rb]: Ah I see, hmm that's a bit weird too as the details for that would just be numbers if I'm understanding that correctly rather than strings 🤔

File: lib/healthcheck/government_data_check.rb (hunk 15 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Nice, thanks, have updated

File: lib/healthcheck/government_data_check.rb (hunk 16 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Yeah i think it's worth it. Have adjusted, didn't realise you could be quite so fluid with durations and numbers.

File: spec/lib/healthcheck/government_data_check_spec.rb (hunk 17 above)
Comment (JonathanHallam) [reply spec/lib/healthcheck/government_data_check_spec.rb]: that seems like a while ago! I've updated the tests to use `travel_to` do you think it's worth having two tests here or would one do, I'm not sure what I'm gaining by testing for 6 and 15 hours...

File: lib/healthcheck/government_data_check.rb (hunk 18 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: It's all weird! I'm trying to test this in integration, I've managed to make it give a warning but not actually raise anything in icinga. This is because of an outstanding puppet change so I'll be looking to deploy puppet to integration in order to test this. I just want to make sure that the messages show up how I expect in Icinga. I'm assuming that we don't really care what it looks like when a person visits `/healthcheck`?

File: spec/requests/healthcheck_spec.rb (hunk 19 above)
Comment (JonathanHallam) [reply spec/requests/healthcheck_spec.rb]: Nice, need to get better with native rspec matchers. Have adjusted.

File: lib/healthcheck/government_data_check.rb [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    def status
+      return GovukHealthcheck::CRITICAL unless government_repo.cache_populated?
+      return GovukHealthcheck::WARNING if government_repo.cache_age > 6.hours
+
+      GovukHealthcheck::OK
+    end
+
+    def message
+      "No government data availible" if status == GovukHealthcheck::CRITICAL
+      warning_details_content if status == GovukHealthcheck:: WARNING
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: This isn't going to work correctly when the status is critical as it isn't the last statement is the second conditional so this will return nil.

You'll need a return statement on the first line or to refactor this a bit. Probably best unit testing it though so this is covered.

File: lib/healthcheck/government_data_check.rb [hunk 21]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: You don't need these anymore since https://github.com/alphagov/content-publisher/pull/1769

File: spec/lib/healthcheck/government_data_check_spec.rb [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    context "when everything is fine" do
+      it "returns ok" do
+        populate_default_government_bulk_data
+        gov_data_check = described_class.new
+        expect(gov_data_check.status).to be(:ok)
+      end
+    end
+  end
+
+  describe "#details" do
```
Comment (kevindew) [inline spec/lib/healthcheck/government_data_check_spec.rb]: Looks like this needs renaming for message

File: spec/lib/healthcheck/government_data_check_spec.rb [hunk 23]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+RSpec.describe Healthcheck::GovernmentDataCheck do
+  describe "#status" do
+    context "when government data is unavailable" do
+      it "returns critical" do
+        gov_data_check = described_class.new
```
Comment (kevindew) [inline spec/lib/healthcheck/government_data_check_spec.rb]: I'd wonder if doing this a bunch is worth the 5 characters we save?

File: lib/healthcheck/government_data_check.rb (hunk 18 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: This was resolved by checking in integration. Turns out that the details method doesn't seem to be shown anywhere. I've replaced it with the "message" method that will show up in Icinga. I've also made changes to the doc to make it clearer what's happening here https://github.com/alphagov/govuk_app_config/pull/135

File: lib/healthcheck/government_data_check.rb (hunk 20 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Absolutely right on this, I didn't realise that's how this worked :/ I've added in unit tests to make sure this gets caught.

File: lib/healthcheck/government_data_check.rb (hunk 21 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Didn't realise that'd happened, nice! Thanks 👍

File: spec/lib/healthcheck/government_data_check_spec.rb (hunk 22 above)
Comment (JonathanHallam) [reply spec/lib/healthcheck/government_data_check_spec.rb]: Should have spotted this, sorry. Fixed.

File: spec/lib/healthcheck/government_data_check_spec.rb (hunk 23 above)
Comment (JonathanHallam) [reply spec/lib/healthcheck/government_data_check_spec.rb]: I've changed this file about a bit and moved this to a `let` block at the top. It's cleaned the file up some especially given the new tests I introduced.

File: lib/healthcheck/government_data_check.rb [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    def status
+      return GovukHealthcheck::CRITICAL unless government_repo.cache_populated?
+      return GovukHealthcheck::WARNING if government_repo.cache_age > 6.hours
+
+      GovukHealthcheck::OK
+    end
+
+    def message
+      return "No government data availible" if status == GovukHealthcheck::CRITICAL
+      return warning_details_content if status == GovukHealthcheck:: WARNING
```
Comment (kevindew) [inline lib/healthcheck/government_data_check.rb]: This last return is redundant

File: spec/requests/healthcheck_spec.rb [hunk 25]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
```
Comment (kevindew) [inline spec/requests/healthcheck_spec.rb]: This frozen string literal was missed.

File: app/controllers/healthcheck_controller.rb [hunk 26]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
```
Comment (kevindew) [inline app/controllers/healthcheck_controller.rb]: Another FSL here too

File: lib/healthcheck/government_data_check.rb (hunk 24 above)
Comment (JonathanHallam) [reply lib/healthcheck/government_data_check.rb]: Yeah, i put it in there to keep the linter happy without a new line but it does look a bit nicer with it, have adjusted

File: spec/requests/healthcheck_spec.rb (hunk 25 above)
Comment (JonathanHallam) [reply spec/requests/healthcheck_spec.rb]: yeah sorry, noticed this immediately after I sent it over, fixed now
